internal/privy/privy.go

```diff
@@ -0,0 +1,213 @@
+	// if so, return error (need a different rkey)
+	_, err := agnostic.RepoGetRecord(ctx, xrpc, "", input.Collection, input.Repo, input.Rkey)
+	if err == nil {
+		return nil, fmt.Errorf("%w: %s", ErrPublicRecordExists, input.Rkey)
+	} else if !errorIsNoRecordFound(err) {
+		return nil, err
+	}
+
+	// If we're here, then this record is to be encrypted and there is no existing public record with rkey
+	// Encrypted -- unpack the request and use special habitat encrypted record lexicon
+	return p.putEncryptedRecord(ctx, xrpc, input.Collection, input.Repo, input.Record, input.Rkey, input.Validate)
+}
+
+type GetRecordResponse struct {
+	Cid   *string `json:"cid"`
+	Uri   string  `json:"uri"`
+	Value any     `json:"value"`
+}
+
+// There are some different scenarios here: (TODO: write tests for all of these scenarios)
+//
+//	1a) cid = that of a non-com.habitat.encryptedRecord --> return that data as-is.
+//	1b) cid = that of a com.habitat.encryptedRecord --> return that data as-is, it will simply be encrypted. getRecord will not attempt to authz and decrypt.
+//	1c) cid = that of a private or public blob --> return that blob as-is.
+//
+// If no cid is provided, fallback to using collection + rkey as the lookup:
+//
+//	2a) collection + rkey = a com.habitat.encryptedRecord --> return that data as-is if exists, which contains a cid pointer to a blob. if no such record exists, return
+//	--) collection + rkey = a non-com.habitat.encryptedRecord:
+//	   2b) if a corresponding record is found, return that
+//	   2c) if no corresponding record is found, attempt to decrypt the record a com.habitat.encryptedRecord would point to for that collection + rkey
+func (p *store) getRecord(ctx context.Context, xrpc *xrpc.Client, cid string, collection string, did string, rkey string) (*agnostic.RepoGetRecord_Output, error) {
+	// Attempt to get a public record corresponding to the Collection + Repo + Rkey.
+	// If the given cid does not point to anything, the GetRecord endpoint returns an error.
+	// Record not found results in an error, as does any other non-200 response from the endpoint.
+	// Cases 1a - 1c are handled directly by this case.
+	output, err := agnostic.RepoGetRecord(ctx, xrpc, cid, collection, did, rkey)
+	// If this is a cid lookup (cases 1a-1c) or the record was found (2a + 2b), simply return ()
+	if err == nil {
+		return output, nil
+	}
+
+	// If the record with the given collection + rkey identifier was not found (case 2c), attempt to get a private record with permissions look up.
+	if strings.Contains(err.Error(), "RecordNotFound") || strings.Contains(err.Error(), "Could not locate record") {
+		return p.getEncryptedRecord(ctx, xrpc, cid, collection, did, rkey)
+	}
+	// Otherwise the lookup failed in some other way, return the error
+	return nil, err
+}
+
+// getEncryptedRecord assumes that the record given by the cid + collection + rkey + did has been encrypted via putRecord and fetches it
```

Comment:
Do we really need both cid and rkey to identify the record? Shouldn't rkey be enough?